processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -267,55 +291,160 @@ static double estimatePredicateSelectivity(
+    }
+  }
+
+  static double estimateSelectivityOfBitmapTree(
       Iterable<ImmutableBitmap> bitmaps,
       long totalNumRows,
       boolean isMultiValueDimension
   )
   {
     long numMatchedRows = 0;
-    final List<ImmutableBitmap> bitmapList = Lists.newArrayList(bitmaps);
-    for (ImmutableBitmap bitmap : bitmapList) {
+    for (ImmutableBitmap bitmap : bitmaps) {
       numMatchedRows += bitmap.size();
     }
 
-    // assume multi-value column if columnCapabilities is null
     if (isMultiValueDimension) {
-      final double estimated = numMatchedRows * Filters.computeNonOverlapRatioFromBitmapSamples(bitmapList)
-                               / totalNumRows;
+      final double estimated = numMatchedRows * computeNonOverlapRatioFromFirstNBitmapSamples(
+          bitmaps
+      ) / totalNumRows;
       return Math.min(1., estimated);
     } else {
       return (double) numMatchedRows / totalNumRows;
     }
   }
 
-  private static double computeNonOverlapRatioFromBitmapSamples(List<ImmutableBitmap> bitmaps)
+  /**
+   * This method is to estimate how many bits of bitmaps are not overlapped in average.
+   * Since a multi-value dimension can have one or more values, one or more bitmaps for that dimension can be set for the same row.
+   * As a result, to get the exact size of unioned bitmaps, which is widely useful for query planning like
+   * filter selectivity estimation, expensive union operations of bitmaps are inevitable.
+   * To avoid such overhead, this method can be used to compute the approximate unioned size based on random sampling.
+   * <p>
+   * The non-overlap ratio can be computed like below.
+   * <p>
+   * nonOverlapRatio(b1, b2) = size(union(b1, b2)) / (size(b1) + size(b2))
+   * <p>
+   * Given bitmaps, this method calculates the non-overlap ratios of N bitmap samples,
+   * and then returns the average of them.
+   *
+   * @param bitmapIndex   bitmap index to retrieve bitmaps
+   * @param bitmapIndexes a list of indexes of bitmaps
+   *
+   * @return approximated average non-overlap ratio of bitmaps
+   *
+   * @see #estimatePredicateSelectivity(ColumnSelector, String, BitmapIndexSelector, Predicate)
+   */
+  private static double computeNonOverlapRatioFromRandomBitmapSamples(
+      BitmapIndex bitmapIndex,
+      List<Integer> bitmapIndexes
+  )
   {
-    final int sampleNum = Math.min(SAMPLE_NUM_FOR_SELECTIVITY_ESTIMATION, bitmaps.size());
+    Preconditions.checkArgument(bitmapIndexes.size() > 0, "empty index list");
+
     double nonOverlapRatioSum = 0.;
+    int sampleNum = Math.min(bitmapIndexes.size(), SAMPLE_NUM_FOR_SELECTIVITY_ESTIMATION);
     for (int i = 0; i < sampleNum; i++) {
-      final ImmutableBitmap b1 = bitmaps.get(random.nextInt(bitmaps.size()));
-      final ImmutableBitmap b2 = bitmaps.get(random.nextInt(bitmaps.size()));
+      final ImmutableBitmap b1 = bitmapIndex.getBitmap(
+          bitmapIndexes.get(ThreadLocalRandom.current().nextInt(bitmapIndexes.size())));
```

Comment:
I suggest to extract `ThreadLocalRandom.current()` as a local variable outside the loop